pkg/controllers/allocation/suite_test.go

```diff
@@ -138,15 +140,18 @@ var _ = Describe("Allocation", func() {
 
 			nodes := &v1.NodeList{}
 			Expect(env.Client.List(ctx, nodes)).To(Succeed())
-			Expect(len(nodes.Items)).To(Equal(6)) // 5 schedulable -> 5 node, 2 coschedulable -> 1 node
 			for _, pod := range schedulable {
 				scheduled := ExpectPodExists(env.Client, pod.GetName(), pod.GetNamespace())
 				ExpectNodeExists(env.Client, scheduled.Spec.NodeName)
 			}
 			for _, pod := range unschedulable {
 				unscheduled := ExpectPodExists(env.Client, pod.GetName(), pod.GetNamespace())
+				if unscheduled.Spec.NodeName != "" {
+					fmt.Printf("This pod shouldn't have been scheduled: %s", pretty.Concise(pod))
```

Comment:
Is the Expect() error message not good enough?

Reply:
I'm not entirely sure if you're being serious or sarcastic, I'll assume you're being serious though. Yes, the Expect() error message is basically worthless. 

```
Allocation Reconcilation [It] should provision nodes for pods with supported node selectors 
/local/workplace2/go/src/karpenter/pkg/controllers/allocation/suite_test.go:109

  Expected
      <string>: thiefcurly
  to equal
      <string>: 

  /local/workplace2/go/src/karpenter/pkg/controllers/allocation/suite_test.go:152
------------------------------
••••
```

That doesn't help me in the slightest other than to tell me that one of the "unschedulable" pods (that term is so confusing, btw, because it means two different things in this test) was scheduled. Once I printed out the pod labels it was trivially easy to debug:

> [1636000414] Controllers/Allocation - 6/6 specs •This pod shouldn't have been scheduled: {"metadata":{"name":"legweak","namespace":"default","selfLink":"/api/v1/namespaces/default/pods/legweak","uid":"013203f3-cf1b-42f1-8580-62112f5c0e0e","resourceVersion":"86","creationTimestamp":"2021-11-04T04:33:55Z","managedFields":[{"manager":"allocation.test","operation":"Update","apiVersion":"v1","time":"2021-11-04T04:33:55Z","fieldsType":"FieldsV1","fieldsV1":{"f:spec":{"f:containers":{"k:{\"name\":\"legweak\"}":{".":{},"f:image":{},"f:imagePullPolicy":{},"f:name":{},"f:resources":{},"f:terminationMessagePath":{},"f:terminationMessagePolicy":{}}},"f:dnsPolicy":{},"f:enableServiceLinks":{},"f:nodeSelector":{".":{},"f:kubernetes.io/os":{}},"f:restartPolicy":{},"f:schedulerName":{},"f:securityContext":{},"f:terminationGracePeriodSeconds":{}},"f:status":{"f:conditions":{".":{},"k:{\"type\":\"PodScheduled\"}":{".":{},"f:lastProbeTime":{},"f:lastTransitionTime":{},"f:reason":{},"f:status":{},"f:type":{}}}}}}]},"spec":{"containers":[{"name":"legweak","image":"k8s.gcr.io/pause","resources":{},"terminationMessagePath":"/dev/termination-log","terminationMessagePolicy":"File","imagePullPolicy":"Always"}],"restartPolicy":"Always","terminationGracePeriodSeconds":30,"dnsPolicy":"ClusterFirst","nodeSelector":{"kubernetes.io/os":"unknown"},"securityContext":{},"schedulerName":"default-scheduler","tolerations":[{"key":"node.kubernetes.io/not-ready","operator":"Exists","effect":"NoExecute","tolerationSeconds":300},{"key":"node.kubernetes.io/unreachable","operator":"Exists","effect":"NoExecute","tolerationSeconds":300}],"priority":0,"enableServiceLinks":true,"preemptionPolicy":"PreemptLowerPriority"},"status":{"phase":"Pending","qosClass":"BestEffort"}}

Reply:
Haha again, as I re-read this it feels inflammatory but I promise I'm being earnest: I truly didn't know if you were being honest or sarcastic, it was not an aggressive message :) To me, the test failures I faced generally gave me back useless messages, I had made an assumption that this was a known issue.

Reply:
I think the error message will be more clear if we say `Expect(unscheduled.Spec.NodeName).To(BeEmpty())`. If the builtin expect matchers are spitting out crappy error messages, there are ways to customize them:
https://github.com/awslabs/karpenter/blob/main/pkg/test/expectations/expectations.go#L57

Reply:
I don't know if I'd consider that alone to be an improvement:

```
Allocation Reconcilation [It] should provision nodes for pods with supported node selectors 
/local/workplace2/go/src/karpenter/pkg/controllers/allocation/suite_test.go:109

  Expected
      <string>: thunderrust
  to be empty

  /local/workplace2/go/src/karpenter/pkg/controllers/allocation/suite_test.go:152
```

I think I'll just make a custom one instead.